game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProPurchaseUtils.java

```diff
@@ -90,7 +91,13 @@ public static boolean canUnitsBePlaced(final List<Unit> units, final PlayerID pl
     }
     final IDelegateBridge bridge = new ProDummyDelegateBridge(ProData.getProAi(), player, data);
     placeDelegate.setDelegateBridgeAndPlayer(bridge);
-    final String s = placeDelegate.canUnitsBePlaced(t, units, player);
+    final String s;
+    if (t.isWater() && Properties.getProduceFightersOnCarriers(data) && units.stream().anyMatch(Matches.unitIsAir())
```

Comment:
This condition is a bit difficult to grok. It would be really nice if it could read something like:
```
return ignoreNewFighters() 
  ? placeDelegate.canUnitsBePlaced(t, CollectionUtils.getMatches(units, Matches.unitIsNotAir()), player) == null
  : placeDelegate.canUnitsBePlaced(t, units, player) == null;

```

Beyond that, I am having some trouble making sense why we check for a water territory, then we prune down to 'not-air', is that us pruning down to naval units?

Reply:
Yeah, in this case 'not air' should be equivalent to 'is sea' since land units can't be placed in sea territories. I felt is was more straightforward to essentially 'remove air' vs 'keep only sea'.

Reply:
Don't want to hold this up, the fix looks legit, but I'm concerned at a potential "yoda" condition here. While it is not unreasonable to figure out what is happening/why we are checking what we are, if we could make it easier/faster to determine the intent behind this code for the sake of faster reading, that would be really helpful. I don't think I'd be able to tell you independent of this PR what the boolean condition was really checking for and why.

Reply:
Added helper method.

Reply:
Perfect, thank you!